### PR TITLE
Wrap request handlers in Ember run loop

### PR DESCRIPTION
**Pull request by alexlafroscia**
- Re-implements `Ember.run` wrapping of handler code, as mentioned here -> https://github.com/ember-cli/ember-ajax/issues/19#issuecomment-169190521
- Adds documentation regarding run loop wrapping, as well as other general cleanup of the Readme

### CODE REVIEW

File: addon/services/ajax.js [hunk 1]
```diff
@@ -113,9 +113,9 @@ export default Ember.Service.extend({
         );
 
         if (response instanceof AjaxError) {
-          reject(response);
+          Ember.run.join(null, reject, response);
```
Comment (toranb) [inline addon/services/ajax.js]: learning opportunity (for me) -why ember.run.join here instead of ember.run?

Comment (rwjblue) [inline addon/services/ajax.js]: `Ember.run` will create a new run loop (even if one is already started), `Ember.run.join` will join an existing run loop or create a new one if one isn't underway.

Comment (toranb) [inline addon/services/ajax.js]: @rwjblue thanks for the TL;DR

My deeper question is - how would this work in the following scenario (run-loop details)

1) user clicks a button w/ onclick
2) that click event is wrapped w/ ember.run (closure action or something lets say)
3) that click event fires this code above
4) when the resolve fires you will most likely not be in that original ember.run (from step 1/2 above) -right?

if you are not as I showed above (total guess -no proof here in code as you can tell) -what is a real use case in which join is successful / useful ?

btw -not trolling here if it sounds that way - just learning a lot about ember.run /when it's used correctly and wanted to learn more :)

**Comment by taras** [comment]
LGTM :+1: to merge

File: README.md [hunk 2]
```diff
@@ -63,12 +64,9 @@ del('/posts/1')                                 // Makes a DELETE request to /po
 
 ### Custom Request Headers
 
-`ember-ajax` allows you to specify headers to be used with a request. This is
-especially helpful when you have a session service that provides an auth token
-that you have to include with the requests to authorize your requests.
+`ember-ajax` allows you to specify headers to be used with a request. This is especially helpful when you have a session service that provides an auth token that you have to include with the requests to authorize your requests.
```
Comment (stefanpenner) [inline README.md]: 80char width please

File: addon/services/ajax.js (hunk 1 above)
Comment (stefanpenner) [inline addon/services/ajax.js]: @toranb runs are intended to be synchronous, so by having an ajax request return asynchronously the run which triggered it should already be complete.

Comment (stefanpenner) [inline addon/services/ajax.js]: On another note though, why are we making this change. `Ember.RSVP` promises are exempt from the `auto-run` assertion, to specifically allow for this sort of case, and prevent introducing accidental ordering differences.

Comment (stefanpenner) [inline addon/services/ajax.js]: I suspect it is to allow for batching of XHR, i suppose if users are doing that they are in power-user land and must already do the right thing....

Comment (taras) [inline addon/services/ajax.js]: @stefanpenner When writing tests that perform async operations, it's useful to be able to easily wait for all promises to resolve before finishing the test. @rwjblue create [`wait()`](https://github.com/switchfly/ember-test-helpers/blob/master/lib/ember-test-helpers/wait.js) for this purpose. This helper relies on the run loop to know if all scheduled timers have been cleared before resolving the `wait()`. Without this, we'd have to track of these promises inside of the service.

Comment (stefanpenner) [inline addon/services/ajax.js]: > @rwjblue create wait() for this purpose. This helper relies on the run loop to know if all scheduled timers have been cleared before resolving the wait()

this isn't correct,  `wait` doesn't detect run-lops rather explicitly detects promises async https://github.com/emberjs/ember.js/blob/master/packages/ember-runtime/lib/ext/rsvp.js#L28

Comment (taras) [inline addon/services/ajax.js]: I referring to this https://github.com/switchfly/ember-test-helpers/blob/master/lib/ember-test-helpers/wait.js#L37

Comment (stefanpenner) [inline addon/services/ajax.js]: Ya, but that is orthogonal, as promises are correctly instrumented to not depend on the legacy polling.

Comment (taras) [inline addon/services/ajax.js]: Ok, @alexlafroscia and I will write a few tests. If we're not able to replicate the problem then we'll just close this issue. Thank you for bringing this to my attention.

**Comment by taras** [comment]
@alexlafroscia pls, hold of on merging this. We need to write some tests to see if this is necessary. 

Ref https://github.com/ember-cli/ember-ajax/pull/40#discussion_r49267568

File: addon/services/ajax.js (hunk 1 above)
Comment (stefanpenner) [inline addon/services/ajax.js]: Well, i think we should actually do this, but we should doit for the right reasons. It exists, to correctly enable batching of XHR. as ed does it https://github.com/emberjs/data/blob/master/addon/adapters/rest.js#L857-L859

Comment (taras) [inline addon/services/ajax.js]: What does `batching of XHR` provide?

**Comment by alexlafroscia** [comment]
@taras with #42 being pulled in, the code that this was based on has changed a lot.  Is it cleaner to just open a new PR? Or try to rebase this branch?

**Comment by taras** [comment]
It'll likely be easier to just open a new PR, but we need to figure out what to with this first.

**Comment by alexlafroscia** [comment]
Alright, I'll hold off on making that then.

File: addon/services/ajax.js (hunk 1 above)
Comment (rwjblue) [inline addon/services/ajax.js]: @stefanpenner - This is required until I make the wait helper in integration tests handle all promises (as you mentioned the wait helper in acceptance tests already does this). The gain is fairly significant (that we can use ember-ajax from component integration tests), and I don't see a significant downside. 

I'm happy to coordinate be general RSVP awareness in the ember-test-helpers wait helper so that we can remove this wrapping, but in the meantime I think we should land it...

Comment (toranb) [inline addon/services/ajax.js]: @stefanpenner @rwjblue thanks for the great rundown (I've learned a ton just reading the comments above). One thing I wanted to clarify quick ... because you are doing a run.join this won't join the original (sync call like a closure action from a button click as you discussed @stefanpenner) but how does this run.join work inside an integration test if I don't wrap my click event (or change event) with ember.run? I ask because as of ember 2.3 I don't wrap my integration clicks/onchanges any longer and I'm curious what the join truly "joins up" with as autorun is disabled in testing

Thank you both in advance -learning 10x from you guys :)

**Comment by taras** [comment]
We're going to go ahead and merge as per https://github.com/ember-cli/ember-ajax/pull/40#discussion_r49268144

We can revisit this in the future it's no longer necessary.
